src/modules/numberplate/index.ts

```diff
@@ -0,0 +1,150 @@
+export type ProvinceObject = Array<{ province: string; codes: Array<number> }>;
+export type plateType = "خودرو" | "موتور سیکلت";
+
+export interface NormilizedPlate {
+	numbers: string;
+	char: string | undefined;
+}
+
+export interface PlateInfoReturnType {
+	template: string;
+	province: Array<string>;
+	type: plateType;
+	category: string | undefined;
+}
+
+class Plate {
+	private plate: NormilizedPlate;
+	private provinceCode!: number;
+	private plateTemplate!: string;
+	private plateType!: plateType;
+	private plateInfo: PlateInfoReturnType;
+
+	constructor(options: PlateOptions) {
+		this.plate = this.normalizePlate(options);
+		this.plateInfo = this.init();
+	}
+
+	private init(): PlateInfoReturnType | never {
+		let plateResult;
+		if (this.plate.numbers?.length === 7) {
+			plateResult = this.initCar();
+		} else if (this.plate.numbers?.length === 8) {
+			plateResult = this.initBike();
+		} else {
+			throw new Error("a Plate must be 7 or 8 digits long");
+		}
+
+		return plateResult;
+	}
+
+	private initCar(): PlateInfoReturnType {
+		this.plateType = "خودرو";
+		this.provinceCode = +this.plate.numbers.slice(5, 7);
+
+		// {first two digits} {plate character} {next three digits} ایران {province code}
+		this.plateTemplate = `${this.plate.numbers.slice(0, 2)}${this.plate.char}${this.plate.numbers.slice(
+			2,
+			5,
+		)}ایران${this.provinceCode}`;
+		const plateCategory = plateTypeJSON.find((t) => t.label === this.plate.char)?.description;
+		const provinces = this.getProvince(carProvinceCodesJSON, this.provinceCode);
+
+		return {
+			category: plateCategory,
+			province: provinces,
+			template: this.plateTemplate,
+			type: this.plateType,
+		};
+	}
+	private initBike(): PlateInfoReturnType {
+		this.plateType = "موتور سیکلت";
+		this.provinceCode = +this.plate.numbers.slice(0, 3);
+
+		// {province code} - {next five digits}
+		this.plateTemplate = `${this.provinceCode}-${this.plate.numbers.slice(3)}`;
+		const provinces = this.getProvince(bikeProvinceCodesJSON, this.provinceCode);
+
+		return {
+			category: undefined, // there is no category on bike plates
+			province: provinces || [],
+			template: this.plateTemplate,
+			type: this.plateType,
+		};
+	}
+
+	private normalizePlate(rawPlate: PlateOptions): NormilizedPlate {
+		let char;
+		let plateNumbers;
+
+		const nonDigitRegex = /\D/g;
+
+		if (typeof rawPlate === "string") {
+			char = rawPlate.match(nonDigitRegex)?.join("");
+			plateNumbers = rawPlate.replace(nonDigitRegex, "");
+		} else {
+			char = rawPlate?.char;
+			plateNumbers = rawPlate.number.replace(nonDigitRegex, "");
+		}
+
+		return {
+			char,
+			numbers: plateNumbers,
+		};
+	}
+
+	private getProvince(Data: ProvinceObject, provinceCode: number) {
+		const provinces = Data.filter((province) => province.codes.find((code) => code === provinceCode));
+		return this.normalizeProvince(provinces);
+	}
+
+	private normalizeProvince(provinces: ProvinceObject) {
```

Comment:
> And my best suggestion is to create a hash-table of provinces and search time complexity in hash-table is `0(1)`, not `o(n)`

yes I wanted to use hash table first but encountered some problems 
1. Not all provinces have unique codes ( like Tehran & Alborz. they both got province code 21). there is a high chance of duplication and rewrites when using hash

2. doesn't hash table creation is always O(n)? like this doesn't matter at the end we gonna have O(n)
because creating is O(n) accessing the value is O(1) result of this will always be O(n)
or I am doing something wrong?

Reply:
one way to solve the duplication and time complexity problem is that i move duplicated codes to their own object
```javascript
const dummy = {
  11: "tehran",
  12: "tehran",
  21: "tehran, alborz",
  .....
}

// access in O(1)
const provinceCode = dummy[code]
```